### PR TITLE
Advanced Atmospheres

### PR DESCRIPTION
**Pull request by Schneegans**
This PR adds the possibility to add alternative atmospheric models to CosmoScout VR. As an example, it adds the [excellent scattering model by Eric Bruneton](https://github.com/ebruneton/precomputed_atmospheric_scattering).

**For reviewing these changes I would suggest to use the GitHub diff only for the files not in `plugins/csp-atmospheres`. All files in `plugins/csp-atmospheres` were nearly rewritten so that it may be easier to look at them as a new plugin.**

The `csp-atmospheres` plugin received a major refactoring. Here are the main consequences:
* The attributes `enableWater`, `waterLevel`, `enableClouds`, and `cloudAltitude` are now stored per planet. Enabling water on Earth will not flood Mars anymore.
* `waterLevel` and `cloudAltitude` are now given in meters, not relative to the atmosphere height anymore.
* `cloudAltitude` can now be configured at runtime in the user interface.
* In non-HDR mode, the atmosphere now performs filmic tonemapping which results in much less over-exposure around the sun during sunset or sunrise.
* The shaders are now loaded from files and not compiled into the binary.
* No support for light shafts anymore. While not impossible, it would be pretty difficult to implement this properly with Eric Bruneton's model given that we use cascaded shadow maps. Furthermore, this feature used to have a terrible performance anyways.
* No graphical tests of the atmosphere anymore. With the new architecture, it's not possible anymore to render an atmosphere without CosmoScout's core classes. We need a new concept for graphical tests anyways...

Other changes:
* `pApproximateSceneBrightness` is now inverted. Before it actually contained something like "approximate scene darkness" :wink:
* The default exposure and glare values as well as the glare-slider mapping have been tweaked for a better appearance of the atmospheres in HDR mode.
* The triangle ordering in the `ToneMappingNode` has been fixed. It now does not need to disable culling.
* The default star rendering mode has been changed to `eSmoothDisc` which looks better in HDR mode (IMHO).

This should be discussed:
* Should HDR mode be enabled per default? It looks better, but also brings some performance issues.

### CODE REVIEW

**Comment by JonasGilg** [comment]
When the sun is over the horizon, but behind some terrain these artifacts come up. 

image

**Comment by JonasGilg** [comment]
On Mars there seems to be a very hard cutoff with HDR enabled:

image

**Comment by Schneegans** [comment]
> On Mars there seems to be a very hard cutoff with HDR enabled:

This is due to the lack of multi-scattering in the Cosmoscout model. This should be the same as in the `main` branch.

**Comment by Schneegans** [comment]
> When the sun is over the horizon, but behind some terrain these artifacts come up. 

This is a limitation of the Bruneton model. It is the result of floating point precision issues. Maybe we can improve this in a custom model in the future!

File: plugins/csp-atmospheres/src/Plugin.cpp [hunk 1]
```diff
@@ -104,85 +104,66 @@ void Plugin::init() {
...
-  });
-
-  mGuiManager->getGui()->registerCallback("atmosphere.setQuality",
-      "Higher values create a more realistic atmosphere.",
-      std::function([this](double value) { mPluginSettings->mQuality = static_cast<int>(value); }));
-  mPluginSettings->mQuality.connectAndTouch(
-      [this](int value) { mGuiManager->setSliderValue("atmosphere.setQuality", value); });
+  mGuiManager->getGui()->registerCallback("atmosphere.setEnableWater",
+      "Enables or disables rendering of a water surface.", std::function([this](bool enable) {
+        if (mActiveAtmosphere != "") {
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/Plugin.cpp:130]:
```suggestion
        if (!mActiveAtmosphere.empty()) {
```

File: plugins/csp-atmospheres/src/Plugin.cpp [hunk 2]
```diff
@@ -104,85 +104,66 @@ void Plugin::init() {
...
-      atmosphere.second->getRenderer().setUseToneMapping(!val, exposure, gamma);
-      if (val) {
-        atmosphere.second->getRenderer().setHDRBuffer(mGraphicsEngine->getHDRBuffer());
-      } else {
-        atmosphere.second->getRenderer().setHDRBuffer(nullptr);
-      }
-    }
-  });
+      "Sets the height of the water surface in meters.", std::function([this](double value) {
+        if (mActiveAtmosphere != "") {
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/Plugin.cpp:139]:
```suggestion
        if (!mActiveAtmosphere.empty()) {
```

File: plugins/csp-atmospheres/src/Plugin.cpp [hunk 3]
```diff
@@ -104,85 +104,66 @@ void Plugin::init() {
...
+      }));
 
-  mAmbientBrightnessConnection =
-      mAllSettings->mGraphics.pAmbientBrightness.connect([this](float val) {
-        for (auto const& atmosphere : mAtmospheres) {
-          float const ambientBrightnessModifier = 0.4F;
-          atmosphere.second->getRenderer().setAmbientBrightness(val * ambientBrightnessModifier);
+  mGuiManager->getGui()->registerCallback("atmosphere.setEnableClouds",
+      "Enables or disables rendering of a cloud layer.", std::function([this](bool enable) {
+        if (mActiveAtmosphere != "") {
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/Plugin.cpp:148]:
```suggestion
        if (!mActiveAtmosphere.empty()) {
```

File: plugins/csp-atmospheres/src/Plugin.cpp [hunk 4]
```diff
@@ -104,85 +104,66 @@ void Plugin::init() {
...
+          auto& settings         = mPluginSettings->mAtmospheres.at(mActiveAtmosphere);
+          settings.mEnableClouds = enable;
+          mAtmospheres.at(mActiveAtmosphere)->configure(settings);
         }
-      });
+      }));
+
+  mGuiManager->getGui()->registerCallback("atmosphere.setCloudAltitude",
+      "Higher values create a more realistic atmosphere.", std::function([this](double value) {
+        if (mActiveAtmosphere != "") {
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/Plugin.cpp:157]:
```suggestion
        if (!mActiveAtmosphere.empty()) {
```

File: plugins/csp-atmospheres/src/models/bruneton/Model.cpp [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+// summed and averaged in each bin (e.g. the value for 360nm is the average
+// of the ASTM G-173 values for all wavelengths between 360 and 370nm).
+// Values in W.m^-2.
+constexpr int    LAMBDA_MIN           = 360;
+constexpr int    LAMBDA_MAX           = 830;
+constexpr double SOLAR_IRRADIANCE[48] = {1.11776, 1.14259, 1.01249, 1.14716, 1.72765, 1.73054,
+    1.6887, 1.61253, 1.91198, 2.03474, 2.02042, 2.02212, 1.93377, 1.95809, 1.91686, 1.8298, 1.8685,
+    1.8931, 1.85149, 1.8504, 1.8341, 1.8345, 1.8147, 1.78158, 1.7533, 1.6965, 1.68194, 1.64654,
+    1.6048, 1.52143, 1.55622, 1.5113, 1.474, 1.4482, 1.41018, 1.36775, 1.34188, 1.31429, 1.28303,
+    1.26758, 1.2367, 1.2082, 1.18737, 1.14683, 1.12362, 1.1058, 1.07124, 1.04992};
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/models/bruneton/Model.cpp:53]:
```suggestion
constexpr std::array<double, 48> SOLAR_IRRADIANCE = {1.11776, 1.14259, 1.01249, 1.14716, 1.72765,
    1.73054, 1.6887, 1.61253, 1.91198, 2.03474, 2.02042, 2.02212, 1.93377, 1.95809, 1.91686, 1.8298,
    1.8685, 1.8931, 1.85149, 1.8504, 1.8341, 1.8345, 1.8147, 1.78158, 1.7533, 1.6965, 1.68194,
    1.64654, 1.6048, 1.52143, 1.55622, 1.5113, 1.474, 1.4482, 1.41018, 1.36775, 1.34188, 1.31429,
    1.28303, 1.26758, 1.2367, 1.2082, 1.18737, 1.14683, 1.12362, 1.1058, 1.07124, 1.04992};
```

File: plugins/csp-atmospheres/src/models/bruneton/Model.cpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+// referencespectra/o3spectra2011/index.html for 233K, summed and averaged in
+// each bin (e.g. the value for 360nm is the average of the original values
+// for all wavelengths between 360 and 370nm). Values in m^2.
+constexpr double OZONE_CROSS_SECTION[48] = {1.18e-27, 2.182e-28, 2.818e-28, 6.636e-28, 1.527e-27,
+    2.763e-27, 5.52e-27, 8.451e-27, 1.582e-26, 2.316e-26, 3.669e-26, 4.924e-26, 7.752e-26,
+    9.016e-26, 1.48e-25, 1.602e-25, 2.139e-25, 2.755e-25, 3.091e-25, 3.5e-25, 4.266e-25, 4.672e-25,
+    4.398e-25, 4.701e-25, 5.019e-25, 4.305e-25, 3.74e-25, 3.215e-25, 2.662e-25, 2.238e-25,
+    1.852e-25, 1.473e-25, 1.209e-25, 9.423e-26, 7.455e-26, 6.566e-26, 5.105e-26, 4.15e-26,
+    4.228e-26, 3.237e-26, 2.451e-26, 2.801e-26, 2.534e-26, 1.624e-26, 1.465e-26, 2.078e-26,
+    1.383e-26, 7.105e-27};
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/models/bruneton/Model.cpp:65]:
```suggestion
constexpr std::array<double, 48> OZONE_CROSS_SECTION = {1.18e-27, 2.182e-28, 2.818e-28, 6.636e-28,
    1.527e-27, 2.763e-27, 5.52e-27, 8.451e-27, 1.582e-26, 2.316e-26, 3.669e-26, 4.924e-26,
    7.752e-26, 9.016e-26, 1.48e-25, 1.602e-25, 2.139e-25, 2.755e-25, 3.091e-25, 3.5e-25, 4.266e-25,
    4.672e-25, 4.398e-25, 4.701e-25, 5.019e-25, 4.305e-25, 3.74e-25, 3.215e-25, 2.662e-25,
    2.238e-25, 1.852e-25, 1.473e-25, 1.209e-25, 9.423e-26, 7.455e-26, 6.566e-26, 5.105e-26,
    4.15e-26, 4.228e-26, 3.237e-26, 2.451e-26, 2.801e-26, 2.534e-26, 1.624e-26, 1.465e-26,
    2.078e-26, 1.383e-26, 7.105e-27};
```

File: plugins/csp-atmospheres/src/models/bruneton/Model.cpp [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  internal::DensityProfileLayer mieLayer(0.0, 1.0, -1.0 / settings.mMieScaleHeight, 0.0, 0.0);
+
+  // Density profile increasing linearly from 0 to 1 between 10 and 25km, and
+  // decreasing linearly from 1 to 0 between 25 and 40km. This is an approximate
+  // profile from http://www.kln.ac.lk/science/Chemistry/Teaching_Resources/
+  // Documents/Introduction%20to%20atmospheric%20chemistry.pdf (page 10).
+  std::vector<internal::DensityProfileLayer> ozoneDensity;
+  ozoneDensity.push_back(
+      internal::DensityProfileLayer(25000.0, 0.0, 0.0, 1.0 / 15000.0, -2.0 / 3.0));
+  ozoneDensity.push_back(internal::DensityProfileLayer(0.0, 0.0, 0.0, -1.0 / 15000.0, 8.0 / 3.0));
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/models/bruneton/Model.cpp:128]:
```suggestion
  ozoneDensity.emplace_back(25000.0, 0.0, 0.0, 1.0 / 15000.0, -2.0 / 3.0);
  ozoneDensity.emplace_back(0.0, 0.0, 0.0, -1.0 / 15000.0, 8.0 / 3.0);
```

File: plugins/csp-atmospheres/src/models/cosmoscout/Model.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+GLuint Model::getShader() const {
+  return mShader.GetFragmentShader();
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+GLuint Model::setUniforms(GLuint program, GLuint startTextureUnit) const {
```
Comment (JonasGilg) [inline plugins/csp-atmospheres/src/models/cosmoscout/Model.cpp:89]:
```suggestion
GLuint Model::setUniforms(GLuint /*program*/, GLuint startTextureUnit) const {
```

**Review by JonasGilg — CHANGES_REQUESTED** [review]

**Review by JonasGilg — APPROVED** [review]
Very nice. I am looking forward to more models being added :)

**Pull request merged**